### PR TITLE
core/vm: check SSTORE access affordability before reading committed state

### PR DESCRIPTION
**Pull request by pratikspatil024**
## Summary

The PIP-88 SSTORE gas function (`makeGasSStoreFuncPIP88`, `core/vm/operations_acl.go`)
performed the committed-state read **before** it knew the slot's access cost. Because
the cold access cost (`ColdSstoreCostPIP88` = 2940) exceeds the reentrancy sentry
(`SstoreSentryGasEIP2200` = 2300), clearing the sentry no longer guarantees the caller
can pay for a cold access — so the read could be performed for a call that then runs out
of gas, i.e. work the gas schedule does not charge for. On bor this path is live behind
Chicago (mainnet, Amoy).

This PR resolves the access cost and warms the slot first, verifies the caller can afford
the access, and only then reads — matching the ordering go-ethereum adopted for its
EIP-8037/8038 SSTORE handler in ethereum/go-ethereum#35261.

**Outcome-neutral.** The guard only rejects callers that already fail the subsequent
charge — the minimum cold total (3040 = 2940 + 100) exceeds the 2940 guard — so no
transaction changes success/failure or gas used; both routes surface `ErrOutOfGas`. Every
returned gas value is byte-identical. Not a hardfork.

Verified on a devnet built from this branch (Chicago active), against the same devnet
built from `develop`:

| | develop | this branch |
| --- | ---: | ---: |
| attack tx `gasUsed` | 454349 | **454349 (identical)** |
| block witness under attack | 21,788 B | **4,243 B (back to baseline)** |

The identical `gasUsed` is the outcome-neutrality check on a live chain — same gas, same
OOG outcome; only the block witness (non-consensus) shrinks, because the reverted read no
longer populates the trie-fetch path that feeds witness collection.

## Executed tests

- `go test ./core/vm/` — new `TestPIP88SStoreAccessCheck` passes (5 boundary cases). It
  uses a read-counting `StateDB` to assert the committed-state read does **not** happen
  once gas is above the sentry but below the access cost — the boundary an error-string
  check alone would miss.
- Full `core/vm` suite unchanged (no gas-value test flips). The one intermittent failure,
  `TestAbortDuringJump/JUMP`, is a pre-existing timing-race flake — it fails identically
  on `develop` with this change stashed out and passes in isolation.
- `go build ./...`, `gofmt`, `go vet ./core/vm/` clean.
- **Devnet (Kurtosis, bor from this branch):** opcode-level `debug_traceCall` sweep on a
  cold slot flips only in the intended window — at 2600 gas the SSTORE now rejects with
  `not enough gas for slot access` before the read, while ≤2300 and ≥2940 are unchanged.
  Baseline-vs-attack witness windows show the witness back at baseline (table above); the
  chain produced blocks normally throughout.

## Rollout notes

- **Not consensus-affecting; no coordinated upgrade required.** Gas charged, transaction
  success/failure, receipts, and state roots are identical before and after — a full-sync
  node produces the same blocks. The only change is a smaller block witness for the
  affected calls.
- **No hardfork.** The fix is a bug fix within the already-active Chicago/PIP-88 behavior,
  not a schedule change.
- **Witness / stateless-sync note for reviewers:** because the fix removes a read, a fixed
  producer emits a slightly smaller witness. Full-sync nodes are unaffected; witness
  producers and stateless-sync consumers should move together during rollout. The witness
  is not committed in the block header on the current (pre-Amsterdam) chain, so there is no
  consensus impact — worth a reviewer confirming against the target release.

### CODE REVIEW

**Comment by pratikspatil024** [comment]
codegenie review

**Comment by pratikspatil024** [comment]
@claude review

File: core/vm/operations_acl.go [hunk 1]
```diff
@@ -102,20 +102,29 @@ func makeGasSStoreFuncPIP88(clearingRefund uint64) gasFunc {
...
 			cost = params.ColdSstoreCostPIP88
+			access = params.ColdSstoreCostPIP88
 			// If the caller cannot afford the cost, this change will be rolled back
 			evm.StateDB.AddSlotToAccessList(contract.Address(), slot)
 		}
+		// Only read the slot once the caller can pay for the access.
+		if contract.Gas < access {
+			return 0, errors.New("not enough gas for slot access")
+		}
+		current, original := evm.StateDB.GetStateAndCommittedState(contract.Address(), slot)
```
[the remark this answers is not in the document]
Comment (pratikspatil024) [reply core/vm/operations_acl.go:127]: This is deliberate. The ordering here — resolve the access cost, warm the slot, check affordability, then read — mirrors the fix go-ethereum applied to the same shape in its EIP-8037/8038 SSTORE handler (ethereum/go-ethereum#35261), which also warms before the affordability check.

As you note, on the reject path the `AddSlotToAccessList` is unwound by the OOG frame revert, so it's outcome-neutral: the only difference from warming-after-the-check is a single map insert that then gets reverted, on a frame that is already OOG'ing. Keeping the upstream ordering for parity, and to avoid splitting the `!slotPresent` block into two. Happy to revisit if we'd prefer to optimize the abort path specifically.

**Review by vbhattaccmu — APPROVED** [review]

**Review by kamuikatsurgi — APPROVED** [review]
